File: data-structures/tree/min_heap.cpp
#include<bits/stdc++.h>
using namespace std;

void min_heapify(vector<int> &arr, int i){
  int n = arr.size();
  int left = 2*i;
  int right = left + 1;
  int smallest = i;
  if(left < n && arr[i] > arr[left]) smallest = left;
  if(right < n && arr[smallest] > arr[right]) smallest = right;

  if(smallest != i){
    swap(arr[i], arr[smallest]);
    min_heapify(arr, smallest);
  }
}

void build_min_heap(vector<int> &arr){
  int n = arr.size();
  for(int i = n/2; i >= 0; i--){
    min_heapify(arr, i);
  }
}

int main(){
  vector<int> arr = {1,4,3,7,8,9,10};
  build_min_heap(arr);
  for(int i: arr) cout << i << " ";
}

File: data-structures/tree/min_heap.py
def min_heapify(list, i):
  n = len(list)
  left = 2*i
  right = left + 1
  smallest = i

  if left < n and list[i] > list[left]: smallest = left
  if right < n and list[smallest] > list[right]: smallest = right

  if smallest != i:
    list[smallest], list[i] = list[i], list[smallest]
    min_heapify(list, smallest)

def build_min_heap(list):
  for i in range(len(list)//2, -1, -1):
    min_heapify(list, i)

list = [1,6,2,5,9,8,7,3,4]
build_min_heap(list)
print(list)
